src/passes/I64ToI32Lowering.cpp

```diff
@@ -0,0 +1,1197 @@
+      highBits,
+      builder->makeGetGlobal(highBitsGlobal, i32)
+    );
+    GetLocal* getLow = builder->makeGetLocal(lowBits, i32);
+    Block* result = builder->blockify(doCall, setHigh, getLow);
+    freeTemp(lowBits);
+    setOutParam(result, highBits);
+    replaceCurrent(result);
+  }
+  void visitCall(Call* curr) {
+    visitGenericCall<Call>(
+      curr,
+      [&](std::vector<Expression*>& args, WasmType ty) {
+        return builder->makeCall(curr->target, args, ty);
+      }
+    );
+  }
+
+  void visitCallImport(CallImport* curr) {
+    // imports cannot contain i64s
+    return;
+  }
+
+  void visitCallIndirect(CallIndirect* curr) {
+    visitGenericCall<CallIndirect>(
+      curr,
+      [&](std::vector<Expression*>& args, WasmType ty) {
+        return builder->makeCallIndirect(
+          curr->fullType,
+          curr->target,
+          args,
+          ty
+        );
+      }
+    );
+  }
+
+  void visitGetLocal(GetLocal* curr) {
+    if (curr->type != i64) return;
+    curr->index = indexMap[curr->index];
+    curr->type = i32;
+    Index highBits = getTemp();
+    SetLocal *setHighBits = builder->makeSetLocal(
+      highBits,
+      builder->makeGetLocal(
+        curr->index + 1,
+        i32
+      )
+    );
+    Block* result = builder->blockify(setHighBits, curr);
+    replaceCurrent(result);
+    setOutParam(result, highBits);
+  }
+
+  void lowerTee(SetLocal* curr) {
+    Index highBits = fetchOutParam(curr->value);
+    Index tmp = getTemp();
+    curr->index = indexMap[curr->index];
+    curr->type = i32;
+    SetLocal* setLow = builder->makeSetLocal(tmp, curr);
+    SetLocal* setHigh = builder->makeSetLocal(
+      curr->index + 1,
+      builder->makeGetLocal(highBits, i32)
+    );
+    GetLocal* getLow = builder->makeGetLocal(tmp, i32);
+    Block* result = builder->blockify(setLow, setHigh, getLow);
+    replaceCurrent(result);
+    setOutParam(result, highBits);
+    freeTemp(tmp);
+  }
+
+  void visitSetLocal(SetLocal* curr) {
+    if (!hasOutParam(curr->value)) return;
+    if (curr->isTee()) {
+      lowerTee(curr);
+      return;
+    }
+    Index highBits = fetchOutParam(curr->value);
+    curr->index = indexMap[curr->index];
+    SetLocal* setHigh = builder->makeSetLocal(
+      curr->index + 1,
+      builder->makeGetLocal(highBits, i32)
+    );
+    Block* result = builder->blockify(curr, setHigh);
+    replaceCurrent(result);
+    freeTemp(highBits);
+  }
+
+  void visitGetGlobal(GetGlobal* curr) {
+    assert(false && "GetGlobal not implemented");
+  }
+
+  void visitSetGlobal(SetGlobal* curr) {
+    assert(false && "SetGlobal not implemented");
+  }
+
+  void visitLoad(Load* curr) {
+    if (curr->type != i64) return;
+    assert(!curr->isAtomic && "atomic load not implemented");
+    Index highBits = getTemp();
+    Index ptrTemp = getTemp();
+    SetLocal* setPtr = builder->makeSetLocal(ptrTemp, curr->ptr);
+    SetLocal* loadHigh;
+    if (curr->bytes > 4) {
```

Comment:
My reading of the spec says that it should be possible to do something like a 6 byte load into an i64. It looks like Binaryen doesn't allow anything but powers of 2 for load sizes, though. Is my reading of the spec wrong?